src/rulesets/Base/indexing.jl

```diff
@@ -2,6 +2,25 @@
 ##### getindex
 #####
 
+function rrule(::typeof(getindex), x::Tuple, ind)
+    y = x[ind]
+    z = map(Returns(NoTangent()), x)
+    project = ProjectTo(x)
+    function getindex_pullback(ȳ, ind::Integer)
+        x̄ = Base.setindex(z, ȳ, ind)
+        return (NoTangent(), project(x̄), NoTangent())
+    end
+    function getindex_pullback(ȳ, ind)
+        x̄ = z
+        for (i, yi) in zip(ind, unthunk(ȳ))
+            x̄ = ntuple(k -> k==i ? x̄[i] + yi : x̄[k], length(z))
+        end
+        return (NoTangent(), project(x̄), NoTangent())
+    end
+    getindex_pullback(ȳ) = getindex_pullback(ȳ, ind)
+    return y, getindex_pullback
+end
```

Comment:
Turns out that `filter(::Tuple)` wants an `rrule` for `getindex` too. I'm not entirely sure this should exist, rather than being left for AD, as it's a pretty low-level operation.

The rrule for Filter could call `rrule_via_ad`. Or perhaps it too need not exist, since that is also just one unrolled `Base.afoldl` call, which AD could surely handle. I added it without enough coffee, did not try to benchmark with/without it.

Reply:
>Or perhaps it too need not exist
"it" here refers to using the `rrule`/`rrule_via_ad`? I don't know what `Base.afoldl` does. Do you mean we could just not use the `rrule`, and write the pullback with `Base.afoldl`? And the "AD could surely handle" refers to higher order derivatives?

Barring any misunderstanding It seems that deleting the rule and using `rrule_via_ad` seems like the most elegant option?

Reply:
Trying things out, the rule for `getindex` is not used. The rule for `filter` does speed this up, although whether it's fair to time Diffractor pre-release I don't know. 
```julia
julia> @btime Diffractor.gradient(x -> +(filter(iseven, x)...), $((1,2,3,4,5)));
  min 18.375 μs, mean 19.967 μs (195 allocations, 9.34 KiB)  # master
  min 5.840 μs, mean 6.101 μs (49 allocations, 1.58 KiB)  # with both rules

julia> @btime Diffractor.gradient(x -> x[1] + x[2], $((1,2,3,4,5)));
  min 367.718 ns, mean 400.458 ns (15 allocations, 368 bytes)  # 
  min 363.527 ns, mean 401.391 ns (15 allocations, 368 bytes)  # does not call the rule
```
The `rrule` for `filter(f, ::Tuple)` is a bit ugly, as it allocates an array, while `filter` is some clever unrolled tuple thing.

Reply:
>Trying things out, the rule for getindex is not used.

Sorry, I am not sure I understand why did it break in the first place, and why is it not breaking now if the rule for `getindex` is not used?

Reply:
These give the right answer with or without rules. If the rule for filter is present it gets called, and it explicitly calls the rule for getindex. But calling getindex does not lead to AD calling the rule for that, according to println, presumably an internal rule handles it before checking CR.

Reply:
Ah, makes sense, thanks. I didn't realise that Tuples worked without the rule :) . Feel free to merge when CI passes.